Fix assert crash "SHOULD NEVER BE REACHED" in StyleRuleCounterStyle

vitorroriz opened the pull request:
#### 37bb9afc0b4754f5f5de0be58acc1bb9752dd12f
<pre>
Fix assert crash &quot;SHOULD NEVER BE REACHED&quot; in StyleRuleCounterStyle
<a href="https://bugs.webkit.org/show_bug.cgi?id=256522">https://bugs.webkit.org/show_bug.cgi?id=256522</a>
rdar://109082260

Reviewed by Antti Koivisto.

We need to add copy for CSSCounterStyleRule since a change
on style will force a copy-on-write operation.

* Source/WebCore/css/CSSCounterStyleRule.h:
Implement copy().

* LayoutTests/http/tests/css/resources/shared.css:
(@counter-style fooCounter):
(@counter-style fooExtends):
- Adding @counter-style rules for exercising the code path tested here.

(@counter-style fooExtends):
* LayoutTests/http/tests/css/shared-stylesheet-mutation-preconstruct-expected.txt:
- Testing mutations on new rules.

* LayoutTests/http/tests/css/shared-stylesheet-mutation-preconstruct-expected.txt:
- Rebasing the this test for new rules.

Canonical link: <a href="https://commits.webkit.org/263870@main">https://commits.webkit.org/263870@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/473f0ddff6318aad2d947707f77d7eac9b2eba4a

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/5948 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/6123 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/6309 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/7501 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/6307 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/5945 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/6342 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/6082 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/8747 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/6056 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/6094 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/5390 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/7561 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/3571 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/5366 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/13296 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/5436 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/5445 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/7666 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/5894 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/4835 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/5326 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/1412 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/9452 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/5687 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

anttijk reviewed (changes requested):
This needs a test. You can expand LayoutTests/http/tests/css/shared-stylesheet-mutation.html or use it as a template.

anttijk reviewed (approved):
You could also expand LayoutTests/http/tests/css/resources/shared-stylesheet-mutation.js to do some `@counter-style` specific mutations.

vitorroriz committed:
Fix assert crash "SHOULD NEVER BE REACHED" in StyleRuleCounterStyle
https://bugs.webkit.org/show_bug.cgi?id=256522
rdar://109082260

Reviewed by Antti Koivisto.

We need to add copy for CSSCounterStyleRule since a change
on style will force a copy-on-write operation.

* Source/WebCore/css/CSSCounterStyleRule.h:
Implement copy().

* LayoutTests/http/tests/css/resources/shared.css:
(@counter-style fooCounter):
(@counter-style fooExtends):
- Adding @counter-style rules for exercising the code path tested here.

(@counter-style fooExtends):
* LayoutTests/http/tests/css/shared-stylesheet-mutation-preconstruct-expected.txt:
- Testing mutations on new rules.

* LayoutTests/http/tests/css/shared-stylesheet-mutation-preconstruct-expected.txt:
- Rebasing the this test for new rules.

Canonical link: https://commits.webkit.org/263870@main
